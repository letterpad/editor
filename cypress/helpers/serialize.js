import { parseFragment, serialize } from "parse5";

function removeData(root) {
  if (root.tagName === "iframe") {
    root.childNodes = [];
  }
  if (root.attrs) {
    root.attrs = root.attrs.filter(
      attr => !attr.name.startsWith("data-") && attr.name !== "class"
    );
  }
  if (root.childNodes) {
    for (const node of root.childNodes) {
      removeData(node);
    }
  }
}

export async function getHtmlContents() {
  return new Promise((resolve, reject) => {
    cy.get("div[contenteditable='true']").then(handle => {
      const fragment = parseFragment(handle[0].innerHTML);
      removeData(fragment);

      resolve(serialize(fragment));
    });
  });
}

export async function textContent(handle) {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
