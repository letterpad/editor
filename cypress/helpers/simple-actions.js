require("cypress-xpath");

export async function repeatKey(key, times) {
  for (let i = 0; i < times; i++) {
    await cy.focused().type(key);
  }
}

export async function clearEditor() {
  return new Promise((resolve, reject) => {
    cy.window().then(win => {
      win.__letterpadEditor.moveToRangeOfDocument();
      cy.focused()
        .type("{backspace}", { force: true })
        .then(resolve);
    });
  });
}

export async function clickXPath(xPath) {
  const ele = cy.xpath(xPath);
  if (ele.length > 0) {
    ele.click({ multiple: true });
  } else {
    ele.click();
  }
}
