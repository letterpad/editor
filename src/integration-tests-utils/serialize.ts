import { ElementHandle } from "puppeteer";
import cheerio from "cheerio";

export type EditorHandle = ElementHandle<Element>;

export async function getHtmlContents(handle: EditorHandle): Promise<string> {
  const context = await handle.executionContext();
  const html = await context.evaluate(node => node.innerHTML, handle);

  const cHtml = cheerio.load(html);
  const allNodes = cHtml("*");
  for (let i = 0; i < allNodes.length; i++) {
    const node = allNodes[i];
    for (let attr in node.attribs) {
      if (attr.startsWith("data-")) {
        allNodes.removeAttr(attr);
      }
    }
  }

  return cHtml.html();
}

export async function textContent(handle: EditorHandle): Promise<any> {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
