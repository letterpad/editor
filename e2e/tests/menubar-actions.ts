import { EditorHandle, getHtmlContents } from "../serialize";
import { clearEditor, repeatKey, clickXPath } from "../simple-actions";
import {
  applyEditorFeatureToSampleText,
  applyEditorFeatureToLine
} from "../compound-actions";
import params from "../params";

describe("features", () => {
  let editorHandle: EditorHandle;

  beforeAll(async () => {
    await page.goto(params.testServer);
    await page.waitForXPath("//div[@contenteditable='true']");
    const handle = await page.$('div[contenteditable="true"]');
    editorHandle = handle!;
    await editorHandle.focus();
  });

  test("clearing editor", async () => {
    await clearEditor(editorHandle);
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("adding some text", async () => {
    await clearEditor(editorHandle);
    await page.keyboard.type("Hello World!");
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("bold", async () => {
    await clearEditor(editorHandle);
    await applyEditorFeatureToSampleText(
      editorHandle,
      "//span[contains(text(), 'format_bold')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("italics", async () => {
    await applyEditorFeatureToSampleText(
      editorHandle,
      "//span[contains(text(), 'format_italic')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("underline", async () => {
    await applyEditorFeatureToSampleText(
      editorHandle,
      "//span[contains(text(), 'format_underline')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("headings", async () => {
    await applyEditorFeatureToLine(
      editorHandle,
      "//span[contains(text(), 'looks_two')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
    await applyEditorFeatureToLine(
      editorHandle,
      "//span[contains(text(), 'looks_3')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("blockquote", async () => {
    await applyEditorFeatureToLine(
      editorHandle,
      "//span[contains(text(), 'format_quote')]"
    );
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  describe("markdown", () => {
    beforeEach(async () => {
      await page.reload();
      await page.waitForXPath("//div[@contenteditable='true']");
      const handle = await page.$('div[contenteditable="true"]');
      editorHandle = handle!;
      await editorHandle.focus();
    });
    test("italic", async () => {
      await clearEditor(editorHandle);
      const text = "foo";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length);
      await page.keyboard.up("Shift");
      await clickXPath("//span[contains(text(), 'format_italic')]");
      const actual = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      await page.keyboard.type("__foo__");
      // because this transformation adds an extra space
      await page.keyboard.press("Backspace");
      const expected = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });

    test("bullet-list", async () => {
      await clearEditor(editorHandle);
      const text = "foo";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length - 2);
      await page.keyboard.up("Shift");
      await clickXPath("//span[contains(text(), 'format_list_bulleted')]");
      const actual = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      await page.keyboard.type("- foo");
      const expected = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });

    test("link", async () => {
      await clearEditor(editorHandle);
      page.evaluate(() => {
        window.prompt = () => "http://google.com";
      });
      const text = "foo";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length);
      await page.keyboard.up("Shift");
      await clickXPath("//span[contains(text(), 'link')]");

      await page.waitFor(10);

      const actual = await getHtmlContents(editorHandle);

      expect(actual).toMatchSnapshot();
    });

    test("bold", async () => {
      await clearEditor(editorHandle);
      const text = "foo";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length);
      await page.keyboard.up("Shift");
      await clickXPath("//span[contains(text(), 'format_bold')]");
      const actual = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      await page.keyboard.type("**foo**");
      // because this transformation adds an extra space
      await page.keyboard.press("Backspace");
      const expected = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });

    test("heading", async () => {
      await clearEditor(editorHandle);
      const text = "bar";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length - 2);
      await page.keyboard.up("Shift");
      await clickXPath("//span[contains(text(), 'looks_two')]");
      const actual = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      await page.keyboard.type("## bar");
      const expected = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });

    test("blockquote", async () => {
      await clickXPath("//span[contains(text(), 'format_quote')]");
      const actual = await getHtmlContents(editorHandle);
      const text = "foo";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length - 2);
      await page.keyboard.up("Shift");

      await clearEditor(editorHandle);
      await page.keyboard.type("> foo");
      const expected = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });

    test("code highlight", async () => {
      await clearEditor(editorHandle);
      const text = "foo";
      await page.keyboard.type(text);
      await page.keyboard.down("Shift");
      await repeatKey("ArrowLeft", text.length);
      await page.keyboard.up("Shift");
      await clickXPath("//span[contains(text(), 'highlight')]");
      const actual = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      await page.keyboard.type("`foo`");
      // because this transformation adds an extra space
      await page.keyboard.press("Backspace");
      const expected = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });
  });
});
