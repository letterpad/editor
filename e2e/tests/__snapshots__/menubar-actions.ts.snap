// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features adding some text 1`] = `"<div style=\\"position: relative;\\"><span><span><span>Hello World!</span></span></span></div>"`;

exports[`features blockquote 1`] = `"<blockquote><span><span><span>These is a sample line of text</span></span></span></blockquote><div style=\\"position: relative;\\"><span><span><span>﻿<br></span></span></span></div>"`;

exports[`features bold 1`] = `"<div style=\\"position: relative;\\"><span><span><span>This is a </span></span><span><strong><span>sample</span></strong></span><span><span> text</span></span></span></div>"`;

exports[`features clearing editor 1`] = `"<div style=\\"position: relative;\\"><span><span><span><span contenteditable=\\"false\\" style=\\"pointer-events: none; display: inline-block; width: 0px; max-width: 100%; white-space: nowrap; opacity: 0.333;\\">Compose a story..</span><span>﻿<br></span></span></span></span></div>"`;

exports[`features headings 1`] = `"<h1><span><span><span>These is a sample line of text</span></span></span></h1><div style=\\"position: relative;\\"><span><span><span>﻿<br></span></span></span></div>"`;

exports[`features headings 2`] = `"<h2><span><span><span>These is a sample line of text</span></span></span></h2><div style=\\"position: relative;\\"><span><span><span>﻿<br></span></span></span></div>"`;

exports[`features italics 1`] = `"<div style=\\"position: relative;\\"><span><span><span>This is a </span></span><span><em><span>sample</span></em></span><span><span> text</span></span></span></div>"`;

exports[`features markdown link 1`] = `"<div style=\\"position: relative;\\"><span><span><span>﻿</span></span></span><a class=\\"link-node\\" href=\\"http://google.com\\"><span><span><span>foo</span></span></span></a><span><span><span>﻿</span></span></span></div>"`;

exports[`features underline 1`] = `"<div style=\\"position: relative;\\"><span><span><span>This is a </span></span><span><u><span>sample</span></u></span><span><span> text</span></span></span></div>"`;
